nservicebus/sagas/saga-not-found.md

```diff
@@ -29,3 +29,74 @@ The ability to provide an implementation for `IHandleSagaNotFound` is especially
+
+## Troubleshooting
+
+*Saga not found* Exceptions can occur when:
+
+1. The saga instance does not exist (yet)
+2. The saga instance has already been removed
+
+It's not always obvious when saga state does or does not exist. Most often the cause is due to race conditions.
+
+Note: Expect out-of-order delivery if the message flow allows it and expect more-than-once processing.
+
+### Out of order
+
+If the design assumes message A will create the saga and message B updates the saga. If message B is received and processed before message A, the saga doesn't exist yet, and the message will be discarded.
+
+Symptoms:
+
+- Sagas are expected to exist but this does not always seems to be the case. Inspecting storage shows that the saga does exist.
+
+Mitigation:
+
+- Use [IAmStartedBy<>](/nservicebus/sagas/#starting-a-saga) instead of `IHandleMessages<>` for **any** message type which can originate from outside of the saga. No matter in which order messages are delivered, any message type processed first must create the saga.
+
+- Each message handler that can start the saga might need to contain logic to check the saga state and see if it is time to take the next action. 
+
+- Messages which are a result of a saga handler do not need this. i.e. A reply to a request made by a saga instance does not need to be able to create a new instance.
+
+### Concurrency
+
+Messages might be processed concurrently which can result in out-of-order processing.
+
+Symptoms:
+
+- Scatter/gather sagas behave unexpectedly and show saga not found exceptions.
+
+Mitigation:
+
+- Reduce processing concurrency to 1 to achieve sequential processing. This only works when only a single endpoint instance is active.
+- Use pessimistic locking on the saga persister, this only works if a storage persister is configured that supports pessimistic locking.
+
+### Initialization
+
+A saga that at creation immediately sends one or more request messages and already is receiving response messages while the handler is still running will result in a response message not able to be correlated if the init handler is still running as the saga state did not yet been persisted
+
+This can happen when:
+
+- Using immediate dispatch
+- Another `IMessageSession` is used instead of `context` or using an alternative protocol
+
+Mitigation:
+
+- Do not immediately send messages at saga creation, first send an init message to ensure the saga state exists. When a message is sent to a component that will respond the saga state will now exist.
+
+### More-than-once processing
+
+A message can be processed more-than-once because it was physically sent multiple times or a transport its consistency results in the same message becoming available again. More-than-once processing on a handler that completes the saga will now result in a saga not found for every duplicate. 
```

Comment:
```suggestion
A message can be processed more than once because it was physically sent multiple times. The same might occur because the transport consistency is "at-least-once," resulting in the same message being consumed more than once. More-than-once processing on a handler that completes the saga will now result in a saga not found for every duplicate. 
```